src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/Dwarf/DwarfBuilder.cs

```diff
@@ -0,0 +1,485 @@
+
+            // Creating a pointer to what DWARF considers Void type (DW_TAG_unspecified_type -
+            // per http://eagercon.com/dwarf/issues/minutes-001017.htm) leads to unhappiness
+            // since debuggers don't really know how to handle that. The Clang symbol parser
+            // in LLDB only handles DW_TAG_unspecified_type if it's named
+            // "nullptr_t" or "decltype(nullptr)".
+            //
+            // We resort to this kludge to generate the exact same debug info for void* that
+            // clang would generate (pointer type with no element type specified).
+            if (pointerDescriptor.ElementType == voidTypeIndex)
+            {
+                _dwarfTypes.Add(new DwarfVoidPtrTypeInfo());
+            }
+            else
+            {
+                _dwarfTypes.Add(new DwarfPointerTypeInfo(pointerDescriptor));
+            }
+
+            return (uint)_dwarfTypes.Count;
+        }
+
+        private uint GetSimpleArrayTypeIndex(uint elementIndex, uint size)
+        {
+            if (_simpleArrayDwarfTypes.TryGetValue((elementIndex, size), out uint index))
+            {
+                return index;
+            }
+
+            _dwarfTypes.Add(new DwarfSimpleArrayTypeInfo(elementIndex, size));
+            uint typeIndex = (uint)_dwarfTypes.Count;
+            _simpleArrayDwarfTypes.Add((elementIndex, size), typeIndex);
+
+            return typeIndex;
+        }
+
+        public uint GetArrayTypeIndex(
+            ClassTypeDescriptor classDescriptor,
+            ArrayTypeDescriptor arrayDescriptor)
+        {
+            // Create corresponding class info
+            ClassTypeDescriptor arrayClassDescriptor = classDescriptor;
+
+            List<DataFieldDescriptor> fieldDescriptors = new();
+            ulong fieldOffset = _targetPointerSize;
+
+            fieldDescriptors.Add(new DataFieldDescriptor
+            {
+                FieldTypeIndex = GetPrimitiveTypeIndex(TypeFlags.Int32),
+                Offset = fieldOffset,
+                Name = "m_NumComponents",
+            });
+            fieldOffset += _targetPointerSize;
+
+            if (arrayDescriptor.IsMultiDimensional != 0)
+            {
+                fieldDescriptors.Add(new DataFieldDescriptor
+                {
+                    FieldTypeIndex = GetSimpleArrayTypeIndex(GetPrimitiveTypeIndex(TypeFlags.Int32), arrayDescriptor.Rank),
+                    Offset = fieldOffset,
+                    Name = "m_Bounds",
+                });
+                fieldOffset += 2u * 4u * (ulong)arrayDescriptor.Rank;
+            }
+
+            fieldDescriptors.Add(new DataFieldDescriptor
+            {
+                FieldTypeIndex = GetSimpleArrayTypeIndex(arrayDescriptor.ElementType, 0),
+                Offset = fieldOffset,
+                Name = "m_Data",
+            });
+
+            // TODO: Verify that this makes sense
+            arrayClassDescriptor.InstanceSize = fieldOffset;
```

Comment:
Re: TODO. It is probably the only value that can be put there, since we do not know the actual array length.

Reply:
On second thought you are right. I was thinking about fixed sized arrays but that was never really propagated to the debugging information, or at least not through this code path.

Reply:
BTW, there is some support in DWARF for specifying variable-length arrays, originally for Fortran. I wonder if LLDB supports this.

Reply:
@SingleAccretion I am interested to learn more about that.

That said, I explicitly tried to avoid fixing non-trivial logic flaws for the debugging information in this PR. I expect to fix some of them later in a follow up PRs. It's hard enough to review this as-is and adding any functional changes would only distract reviewers for the original intent and delay the merge.

I'll update the comment for now. If you have any link to the DWARF variable-length array documentation I'd be happy to include a link in the comment.

Reply:
Right, I am only mentioning this since I didn't know if you knew.

> I am interested to learn more about that.

The general idea is to use an expression for `DW_AT_upper_bound`/`DW_AT_count`. In the copy of the DWARF5 standard I have, there is a nice example in `D.2.1 Fortran Simple Array Example`.